api4/bot.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package api4
+
+import (
+	"net/http"
+
+	"github.com/mattermost/mattermost-server/model"
+)
+
+func (api *API) InitBot() {
+	api.BaseRoutes.Bots.Handle("", api.ApiHandler(createBot)).Methods("POST")
+	api.BaseRoutes.Bot.Handle("", api.ApiSessionRequired(patchBot)).Methods("PUT")
+	api.BaseRoutes.Bot.Handle("", api.ApiSessionRequired(getBot)).Methods("GET")
+	api.BaseRoutes.Bots.Handle("", api.ApiSessionRequired(getBots)).Methods("GET")
+	api.BaseRoutes.Bot.Handle("/disable", api.ApiSessionRequired(disableBot)).Methods("POST")
+}
+
+func sessionHasPermissionToManageBot(c *Context, userId string) *model.AppError {
+	existingBot, err := c.App.GetBot(c.Params.UserId, true)
+	if err != nil {
+		return err
+	}
+
+	if existingBot.CreatorId == c.App.Session.UserId {
+		if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_BOTS) {
+			return c.MakePermissionError(model.PERMISSION_MANAGE_BOTS)
+		}
+	} else {
+		if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_MANAGE_OTHERS_BOTS) {
+			return c.MakePermissionError(model.PERMISSION_MANAGE_OTHERS_BOTS)
+		}
+	}
+
+	return nil
+}
+
+func createBot(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_CREATE_BOT) {
+		c.SetPermissionError(model.PERMISSION_CREATE_BOT)
+		return
+	}
+
+	botPatch := model.BotPatchFromJson(r.Body)
+	if botPatch == nil {
+		c.SetInvalidParam("bot")
+		return
+	}
+
+	bot := &model.Bot{
+		CreatorId: c.App.Session.UserId,
+	}
+	bot.Patch(botPatch)
+
+	createdBot, err := c.App.CreateBot(bot)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	w.WriteHeader(http.StatusCreated)
+	w.Write(createdBot.ToJson())
+}
+
+func patchBot(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.RequireUserId()
+	if c.Err != nil {
+		return
+	}
+
+	if err := sessionHasPermissionToManageBot(c, c.Params.UserId); err != nil {
+		c.Err = err
+		return
+	}
+
+	botPatch := model.BotPatchFromJson(r.Body)
+	if botPatch == nil {
+		c.SetInvalidParam("bot")
+		return
+	}
+
+	updatedBot, err := c.App.PatchBot(c.Params.UserId, botPatch)
+	if err != nil {
+		c.Err = err
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+	w.Write(updatedBot.ToJson())
+}
+
+func getBot(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !c.App.SessionHasPermissionTo(c.App.Session, model.PERMISSION_READ_BOTS) {
```

Comment:
@jwilander, I realized later that a `READ_OTHERS_BOTS` would require a bunch of store work to filter out bots that weren't created by me. Since the extra granularity isn't cheap (and wasn't part of the original spec), I figured I'd leave it out.

Reply:
I woke up this morning and realized I over-estimated the work here, so I went ahead and split out a `READ_OTHERS_BOTS`. I also changed semantics so that `MANAGE_OTHERS_BOTS` implies `MANAGE_BOTS` (why would you be able to manage others but not your own), matching the new `READ_OTHERS_BOTS` and `READ_BOTS` parameters.